lesson16/main.js

```diff
@@ -0,0 +1,111 @@
+// json = https://mocki.io/v1/d3176c84-479a-47a3-a8b6-7d597b55d8b4
+const ul = document.getElementById("js-ul");
+const li = document.createElement("li");
+const contetns = document.getElementsByClassName("tab-content");
+const tabs = document.getElementsByClassName("tab-item");
+
+const displayInfo = (info) => {
+  const li = document.createElement("li");
+  if (typeof info === "string") {
+    li.textContent = info;
+    ul.appendChild(li);
+    return console.log(info);
+  }
+  throw new Error("Please pass a string value");
+};
+
+const displayErrorInfo = (error) => {
+  console.log(error);
+  li.textContent = `Failed to get data from server`;
+  ul.appendChild(li);
+};
+
+const getData = async (api) => {
+  let response = null;
+  try {
+    response = await fetch(api);
+  } catch (error) {
+    displayErrorInfo(error);
+  }
+  if (response.ok) {
+    return response.json();
+  } else {
+    displayErrorInfo(response);
+  }
+};
+
+const checkResponse = async (callBackFunction) => {
+  let response = await getData(
+    "https://mocki.io/v1/40cbd0f7-2a64-4b95-a991-57ff982901a8"
+  );
+  if (response.data) {
+    callBackFunction(response.data);
+  } else {
+    displayInfo("no data available");
+  }
+};
```

Comment:
Thank you Mai san for your feedback
I was not thinking much that api was called twice until you pointed it out.
The suggested code is a lot more easier to read and understand

Also it was really good to know that I do not have to call createTabs and createArticle functions inside of the displayTabsAndContents function in order to wait for them to be rendered first, and it works in the suggested way.

It was a great take in!

Reply:
Thanks for the correction!
The error handling here is assumed to be as follows. I hope this is helpful. I thought it would be fine for the next implementation, but I will comment on it.

```JavaScript
const init = async () => {
  showLoading();
  try{
      const newsGenres = await checkResponse();
      if (!newsGenres) {
        return; // early return;
    }
    createTabs(newsGenres);
    createArticles(newsGenres);
    displayTabsAndContents();
  }catch(error){
    console.error(error);
  }finally{
    removeLoading();
  }
}
```

Reply:
thank you for the suggestion Mai-san
Moarita-san also mentioned that part.
So I added the process for false case
Please check the commit
https://github.com/Saki-Ota/JS_practice/commit/578cec1800b70a279d2ab3b12b836c1b9dc37f13